Add SSL context for RedisCluster connections

robin-brabants opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
This small feature allows to pass an [sslContext](https://www.php.net/manual/en/context.ssl.php) to the [RedisCluster constructor](https://github.com/phpredis/phpredis/blob/develop/cluster.md) from the phpredis extension.

Without this feature, one could not connect to a cluster using TLS. Now when specifying the correct protocol for the seed nodes (e.g.: 'seeds' => ['tls://nodeId-01.xxx:6379', 'tls://nodeId-02.xxx:6379']) and passing an sslContext one can connect to a cluster using TLS .

One can pass an empty array for the sslContext, but also explicitly specify non-default options for the TLS connection. For a full list of ssl context options see: https://www.php.net/manual/en/context.ssl.php.

This for example enables one to connect to an AWS elasticache redis cluster with in-transit encryption set to required due to security reasons.


<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Ocramius commented:
> TARGET THE develop BRANCH????

Is the template still foobar'd ?

Ocramius commented on src/RedisClusterOptions.php:
```diff
@@ -222,4 +224,14 @@ public function setPassword(string $password): void
         $this->password = $password;
     }
+
+    public function getSslContext(): array
+    {
+        return $this->sslContext;
```
Is there any way to document this better? :thinking: 

at least `array<string, mixed>`, I suppose? Or `array{verify_peer: bool, ...array<string, mixed>}`?

robin-brabants replied:
Yeah it makes sense to document the field and getter return type as array<string, mixed>.

The single source of truth for recognized options in this array would be: https://www.php.net/manual/en/context.ssl.php. Would it be useful to put this in the docstring?

Ocramius replied:
Before throwing effort at it, let's wait for @boesing, who's the main maintainer here :) 

Want to see if we should support the whole SSL context, or just parts of it.

boesing replied:
Oh yah, thats exactly what I requested below.
I'd say lets only return an object, that would allow us to consume that context in a more OOP way. I'd like to prevent arrays as much as possible. IMO we should allow an array shape and an object during instantiation, we could provide a proper Options array shape on class-level, I might want to require that in the next major version anyways for *every* options of an adapter. That would basically allow us to have proper type-safety via static analysis and allows us to have an object instantiated from an array.

Since `laminas-cache` depends on `laminas-stdlib`, we could even implement https://github.com/laminas/laminas-stdlib/blob/ec5f3d3f1a2c44575d3369981ff5a1058aacf77d/src/ArraySerializableInterface.php to have a method to pass the SSL context as an array to the `RedisCluster#__construct`.

WDYT? 

----

I stumbled over `template-type` annotation in phpstan and raised a feature request (which I might implement on my own at some point) in psalm: https://github.com/vimeo/psalm/issues/10849

Ocramius reviewed (commented):
I understand where this comes from, but the use case of `verify_peer => false` seems kinda anachronistic nowadays.

Possibly better to provision both host and client with a dummy certificate via https://github.com/FiloSottile/mkcert/blob/2a46726cebac0ff4e1f133d90b4e4c42f1edf44a/README.md ?

robin-brabants commented:
> > TARGET THE develop BRANCH????
> 
> Is the template still foobar'd ?

The template mentions to make a PR for features against a development branch:

Pick the target branch based on the following criteria:
  * New feature, or refactor of existing code: develop branch

robin-brabants commented:
> I understand where this comes from, but the use case of `verify_peer => false` seems kinda anachronistic nowadays.
> 
> Possibly better to provision both host and client with a dummy certificate via https://github.com/FiloSottile/mkcert/blob/2a46726cebac0ff4e1f133d90b4e4c42f1edf44a/README.md ?

it is just an example for a unit test

robin-brabants commented on composer.json:
```diff
@@ -8,7 +8,7 @@
     "require": {
         "php": "~8.1.0 || ~8.2.0 || ~8.3.0",
-        "ext-redis": "^5.0.2 || ^6.0",
+        "ext-redis": "^5.3.2 || ^6.0",
```
Passing a ssl/tls context to the RedisCluster constructor from the phpredis extension is apparently only introduced in version 5.3.2: https://github.com/phpredis/phpredis/blob/develop/CHANGELOG.md 
I noticed this due to the failing pipeline.
This would of course be a breaking change though.

boesing replied:
I havent had this kind of change ever but I remember that I did modified some redis range in the past. IMO having a check during CD that all the required extension versions are fulfilled should be a good thing, so I do not have a problem with that change personally, but how is this kind of stuff handled in other OSS.
Do you have experience here @Ocramius?

robin-brabants commented:
I added a few commits which should fix the failing integration tests. Could they be ran again?

boesing commented on psalm.xml:
```diff
@@ -50,6 +50,15 @@
                 <directory name="*"/>
             </errorLevel>
         </RedundantCastGivenDocblockType>
+        <!--
```
I'd put this into the code via `@psalm-suppress` as that would be reported once its fixed in upstream and thus we are able to remove that suppression.
Suppressions in the psalm config are not reported and thus would suppress this issue for like forever which might introduce problems somewhen in the future.

Would you mind create an issue in psalm as well and/or a PR to update the signature so that we can remove the inline `@psalm-suppression` at some point?
That would be decent value.

robin-brabants replied:
Makes sense, I changed this. Also made a PR for changing this in psalm and referenced it in the docstring: https://github.com/vimeo/psalm/pull/10862

boesing commented on src/RedisClusterOptions.php:
```diff
@@ -53,6 +53,9 @@ final class RedisClusterOptions extends AdapterOptions
 
     private string $password = '';
 
+    /** @psalm-var array<non-empty-string,mixed>|null */
```
do we know what context exists?
Can we probably have an object instead of an array?

boesing commented on test/unit/RedisClusterOptionsTest.php:
```diff
@@ -46,6 +47,7 @@ public function testCanHandleOptionsWithNodename(): void
         $this->assertEquals(false, $options->isPersistent());
         $this->assertEquals('1.0', $options->getRedisVersion());
         $this->assertEquals('secret', $options->getPassword());
+        $this->assertEquals(['verify_peer' => false], $options->getSslContext());
```
I'd rather see an object here.
These are supported SSL context entries: https://www.php.net/manual/en/context.ssl.php

boesing commented on test/unit/RedisClusterOptionsTest.php:
```diff
@@ -38,6 +38,7 @@ public function testCanHandleOptionsWithNodename(): void
             'persistent'    => false,
             'redis_version' => '1.0',
             'password'      => 'secret',
+            'ssl_context'   => ['verify_peer' => false],
```
Lets allow instantiation from array shape and object (which needs to be implemented).
Allowing an array with whatever key might introduce problems if keys are passed as `ssl_context` which are not allowed. If we can prevent that from static analysis level, I'd say lets do so.

boesing commented on test/unit/RedisClusterOptionsTest.php:
```diff
@@ -65,6 +68,7 @@ public function testCanHandleOptionsWithSeeds(): void
         $this->assertEquals(false, $options->isPersistent());
         $this->assertEquals('1.0', $options->getRedisVersion());
         $this->assertEquals('secret', $options->getPassword());
+        $this->assertEquals(['verify_peer' => false], $options->getSslContext());
```
https://github.com/laminas/laminas-cache-storage-adapter-redis/pull/85/files#r1534542474

boesing reviewed (changes requested):
I'd added some comments, would be decent if we can improve this a little bit.
Sorry for the late response, I am not doing too much OSS at the moment.

boesing commented:
I wonder if we can add an integration test where we actually setup a redis cluster with SSL and self-signed certificate to verify if this is actually working. I don't expect this in this PR tho as I think it would already suffice and I can understand that taking time to implement this kind of setup would be too much.
It took me quite some time to even have redis cluster docker setup up and running, since the docker image we are using here does not support SSL right now, we would have to create an own docker image. Thus I think I will create an issue off this comment once we merged this PR.

robin-brabants commented:
Implemented an sslContext class right now which contains the [sslcontext](https://www.php.net/manual/en/context.ssl.php) options in the fields. 

I omitted the 'capture_peer_cert' and 'capture_peer_cert_chain' options since they are not relevant for us in my opinion.

I wanted to make some of the default values clearly visible in the constructor. I made them correspond to the actual sslContext default options at least for more recent php versions with the exception of 'SNI_enabled' which would make sense in my opinion to enable by default if possible.

robin-brabants commented:
@Ocramius @boesing 
Are there still any changes required for this PR

boesing commented:
Sorry, havent reviewed yet. Will probably do so either today or tomorrow evening.

boesing committed:
support passing ssl context for RedisCluster

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
raise minimum required version for the phpredis extension since the ssl/tls context parameter was introduced in version 5.3.2

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
improve type documentation of the redis cluster ssl context

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
set the redis cluster ssl context option default to null, otherwise a TLS connection is attempted by default

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
suppress psalm validation errors for the setter setSslContext and the phpredis RedisCluster class

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
implement an sslContext class which can be used for passing the sslContext to the RedisCluster

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
move psalm validation error suppression for the RedisCluster constructor to the code

Signed-off-by: robin-brabants <robin.brabants@tngtech.com>
Co-authored-by: robin-brabants <robin.brabants@tngtech.com>
Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: `SslContext` and its instantiation

This refactors some of the `SslContext` implementation to have better readability, less complexity (regarding camel case handling, etc.) and explicit context serialization.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
I refactored the code a little bit, especially since I am not a fan from those abbreviations all round tech.
I renamed some properties but still preserve the `context` notation when converting the object to the `ssl context` array.

I'd say we should only allow 1:1 array notations which would be accepted by PHP as well, thats why I removed all that camel case handling and just have two dedicated methods, a named factory to instantiate our object from an SSL context array (in the notation supported by PHP) and a method which converts the object back to a PHP supported SSL context array.

I also do not allow `RedisClusterOptions#setSslContext` to consume an array. IMHO if that is being called from somewhere, it should be either `null` or the `SslContext` object.
I did override `AdapterOptions#setFromArray` to ensure that `ssl_context` or `sslContext` array keys are always converted to an object in case an array is passed.
So test is still passing.

I also removed all the ssl context defaults. That prevents us from having issues with several PHP versions where (whyever) defaults might change and then we have to see how to provide defaults for both versions. I do not really see that we should do that and thus I'd say lets allow passing `null` for any option and in case it is not explicitly set (i.e. `null`), we do not pass that in the `array` we pass to `RedisCluster#__construct`.

I also would say we should not require `ext-openssl` since it would be only required from upstream projects as this component does pretty much work without that extension (as I removed all the defaults which were using `OPENSSL_*` constants) and thus I'd say, that is a requirement which should be added by projects rather than this library.

----

I'd love to get some review of my changes and some feedback. If that works for you and you still see the feature is serving your needs, I'd love to get a quick 👍🏼  so that we can proceed. 
Since I do not have a cluster serving via TLS, I am unable to verify if everything is working. I've created #88 to keep track that we should have integration tests up and running so that we do not accidentally break stuff.

But I'd be fine with releasing this without actual integration tests as its a new feature which needs to be explicitly implemented by projects and thus its okayish to depend on projects having integration testings up and running until we implement #88.

Thanks for your massive work here and I'm happy to see your feedback.

----

_Had to rebase with latest 2.8.x as renovate updated `composer.lock`..._

robin-brabants commented:
Sorry that I only see your improvement right now, but was a bit preoccupied.
I would have some time again after the weekend and will give my feedback on Monday.

robin-brabants commented on src/SslContext.php:
```diff
@@ -4,197 +4,217 @@
-
-use const OPENSSL_DEFAULT_STREAM_CIPHERS;
-use const OPENSSL_TLSEXT_SERVER_NAME;
+use SensitiveParameter;
```
nice addition, but it is only supported from php 8.2 onwards

boesing replied:
Yah, it has no impact on PHP 8.1 but there is not an issue either (unless some1 parses attributes and tries to instantiate them). I think this is okay but still good find.

robin-brabants commented on test/unit/SslContextTest.php:
maybe it is overkill, but we could include some testcase for the  'toSslContextArray' instance method if we provide some non-null parameters to the constructor

boesing replied:
Its not explicitly tested but implicitly via the array notations.
I'd be fine with that for now, that does already verify if expected properties are set with expected values and that would be good enough for me.
Lets ship, we can still add more QA stuff in future releases but for now we still have 100% coverage.

robin-brabants commented:
@boesing 
Your changes look great to me! I just had two minor comments.
For me it is fine to merge this PR as is.

boesing commented:
Thanks @robin-brabants!
